docs: ADR-106: gRPC API

thanethomson opened the pull request:
Partially addresses #81.

[:book: Rendered](https://github.com/cometbft/cometbft/blob/thane/81-grpc-adr/docs/references/architecture/adr-106-grpc-api.md)

There's already a partial implementation of this on the `feature/adr101-pull-companion` branch. See #818.

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

lasarojc commented on docs/architecture/adr-105-grpc-api.md:
```diff
@@ -0,0 +1,222 @@
+  `GetVersion` method needs to be quite stable.
+  - `GetVersion` - Query the version of the software and protocols employed by
+    the node (e.g. CometBFT, ABCI, block, P2P and application version).
+- `NodeService` - Provides information about the node providing the gRPC
```
could the `NodeService` run on a machine different from the one running the full-node?
If so, is the information returned here from the node serving gRPC or the full node or both?

thanethomson replied:
No, Protobuf "services" are just logical ways of grouping sets of RPC methods. These services are always exposed from a full node or a validator.

thanethomson replied:
Ah, I see. If the RPC companion had to also provide the same gRPC interface, it would proxy such methods. There are only a handful of methods where it would pull its data from a separate database, e.g. fetching blocks, block results, historical transactions, etc.

But methods like getting node status, performing an ABCI query, etc. have to be proxied through to the upstream full node to which the companion is attached.

@andynog is looking into this in ADR 102.

thanethomson replied:
Nothing stops the companion from exposing _additional_ services that are specific to the companion itself. But I don't think it should modify or expand on any of the data or services it proxies.

lasarojc commented on docs/architecture/adr-105-grpc-api.md:
```diff
@@ -0,0 +1,222 @@
+  created, and all old versions would need to be maintained, since the
+  `GetVersion` method needs to be quite stable.
+  - `GetVersion` - Query the version of the software and protocols employed by
+    the node (e.g. CometBFT, ABCI, block, P2P and application version).
```
could add "version of any companions" to the list, even though it does not need to be exhaustive.

thanethomson replied:
I wouldn't. That would involve developing some more complex handshake mechanism with the data companion, for which I don't see a need at the moment.

romac commented:
Looks great!

romac commented:
> Glad to hear you think so! Are there any other endpoints that you'd like to see exposed here @romac?

Here's the full list of RPC endpoints we use from Hermes: https://hedgedoc.informal.systems/FmNN0xuESR2wS1ZoRhuU8g

Aside from the first two endpoints which are strictly required for relaying (`/health` & `/consensus_params`), the gRPC list is missing these three endpoints:

- `/header`
- `/tx_search`
- `/block_search`

While it would be nice to get rid of JSON-RPC altogether, since we'll have to maintain multi-version support for some time we could initially do without and keep hitting the JSON-RPC endpoints for those, and wait until they are added at a later time.

thanethomson commented:
@hdevalence I've added a `NetworkService` from which you can query the genesis data and currently connected peers - does that suffice? Do you need the listener info that the `/net_info` provides?

@romac I've updated the endpoints - will this suffice for the relayer?

romac commented:
> @romac I've updated the endpoints - will this suffice for the relayer?

I might be missing something but I believe the `/latest_commit`, `/commit`, and `/validators` endpoints are missing. Those are required by the light client to fetch and verify light blocks.

Aside from that, looks great! 🙏

tony-iqlusion commented:
@thanethomson what about privval support? (previously tracked as https://github.com/tendermint/tendermint/issues/9256)

If that's in the cards, I have some feedback on the design of that API based on some recent experiences (namely it would be nice if there were an error which KMS systems can use to communicate the last signed h/r/s back to CometBFT in the event of double signing, so CometBFT could potentially update its round counter in response)

thanethomson commented:
> what about privval support?

That could definitely be on the cards @tony-iqlusion. We do have a separate tracking issue for that in #476, but we'd be very interested to hear what your ideal API/interaction model would look like.

tony-iqlusion commented:
Really all I'd like is a documented standard way of communicating a double signing error from `sign_vote` and `sign_proposal` and including metadata that CometBFT can parse and use to inform future signing requests. Something like the following:

- On double signing error, return 409 Conflict to indicate a signature at the requested height already exists
- The error should include metadata for `last_height` in `h/r/s` format, which reflects the last height a signature was computed for

If `last_height` reflects the current block height, but CometBFT's round counter is lower than the KMS's, then CometBFT can advance the round counter to one more than the one included in `last_height`.

Right now CometBFT does not handle `privval` errors very well. This was pretty painful with the last Neutron halt, where the chain halted for several hours and validators needed to resync. When the halt height was reached syncing, it took an additional hour for validators using TMKMS to reach the correct round to participate in consensus. It seems settings like `timeout_prevote(_delta)` and `timeout_precommit(_delta)` are not properly honored in this case (which perhaps I should report as a separate bug).

melekes commented:
@thanethomson what's the state of this work? Should we take over here?

thanethomson commented:
> @thanethomson what's the state of this work? Should we take over here?

Yes please @melekes!

adizere reviewed (approved):
Pre-approving based on a sync review with Andy today.

What is left:
* double-check comments from Henry & Tony & Romain
* validate the Go API server instantiation
* double-check potential name conflict "ADR 105"
